plugins/chanbackup.c

```diff
@@ -0,0 +1,393 @@
+{
+
+	size_t i;
+	const jsmntok_t *t;
+	struct json_stream *response;
+
+	response = jsonrpc_stream_success(cmd);
+
+	json_for_each_obj(i, t, params)
+		json_add_tok(response, json_strdup(tmpctx, buf, t), t+1, buf);
+
+	return command_finished(cmd, response);
+}
+
+/* Recovers the channels by making RPC to `recoverchannel` */
+static struct command_result *recover(struct command *cmd,
+				      const char *buf,
+                                      const jsmntok_t *params)
+{
+	struct out_req *req;
+	u64 version;
+	u32 timestamp;
+	struct scb_chan **scb;
+
+	if (!param(cmd, buf, params, NULL))
+		return command_param_failed();
+
+	u8 *res = decrypt_scb(cmd->plugin);
+
+	if (!fromwire_static_chan_backup(cmd,
+                                         res,
+                                         &version,
+                                         &timestamp,
+                                         &scb)) {
+		plugin_err(cmd->plugin, "Corrupted SCB!");
+	}
+
+	if (version != VERSION) {
+		plugin_err(cmd->plugin,
+                           "Incompatible version, Contact the admin!");
+	}
+
+	req = jsonrpc_request_start(cmd->plugin, cmd, "recoverchannel",
+				after_recover_rpc,
+				&forward_error, NULL);
+
+	json_array_start(req->js, "scb");
+	for (size_t i=0; i<tal_count(scb); i++) {
+		u8 *scb_hex = tal_arr(cmd, u8, 0);
+		towire_scb_chan(&scb_hex,scb[i]);
+		json_add_hex(req->js, NULL, scb_hex, tal_bytelen(scb_hex));
+	}
+	json_array_end(req->js);
+
+	return send_outreq(cmd->plugin, req);
+}
+
+static void update_scb(struct plugin *p, const char *scb_buf)
+{
+
+	/* If the temp file existed before, remove it */
+	unlink_noerr("scb.tmp");
+
+	int fd = open("scb.tmp", O_CREAT|O_EXCL|O_WRONLY, 0400);
+	if (fd<0)
+		plugin_err(p, "Opening: %s", strerror(errno));
+
+	plugin_log(p, LOG_DBG, "Updating the SCB file...");
+
+	write_hsm(p, fd, scb_buf);
+
+	/*~ fsync (mostly!) ensures that the file has reached the disk. */
+	if (fsync(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+
+	/*~ This should never fail if fsync succeeded.  But paranoia good, and
+	 * bugs exist. */
+	if (close(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+	/*~ We actually need to sync the *directory itself* to make sure the
+	 * file exists!  You're only allowed to open directories read-only in
+	 * modern Unix though. */
+	fd = open(".", O_RDONLY);
+	if (fd < 0) {
+		plugin_log(p, LOG_DBG, "Opening: %s", strerror(errno));
+	}
+	if (fsync(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+	close(fd);
+	unlink_noerr("emergency.recover");
+	rename("scb.tmp", "emergency.recover");
+}
+
+static struct command_result *after_staticbackup(struct command *cmd,
+					         const char *buf,
+					         const jsmntok_t *params,
+					         void *cb_arg UNUSED)
+{
+	const jsmntok_t *scbs = json_get_member(buf, params, "scb");
+	plugin_log(cmd->plugin, LOG_INFORM, "Updating the SCB");
+
+	update_scb(cmd->plugin, json_strdup(tmpctx, buf, scbs));
+	return notification_handled(cmd);
```

Comment:
Hand this the buf and scbs token, and have it iterate through, rather than covnerting to a string and back...

This is neater, since scbs->size will tell it immediately how many scbs there are.